src/doc/src/guide/build-performance.md

```diff
@@ -0,0 +1,76 @@
+
+Example workflows to consider include:
+- Compiler feedback as you develop (`cargo check` after making a code change)
+- Test feedback as you develop (`cargo test` after making a code change)
+- CI builds
+
+## Cargo and Compiler Configuration
+
+> Note that some approaches described in this section currently require using the nightly toolchain.
+
+Cargo uses configuration defaults that try to balance several aspects, including debuggability, runtime performance, build performance, binary size and others. This section describes several approaches for changing these defaults that should be designed to maximize build performance.
+
+You can set the described options either in the [`Cargo.toml` manifest](../reference/profiles.md), which will make them available for all developers who work on the given crate/project, or in the [`config.toml` configuration file](../reference/config.md), where you can apply them only for you or even globally for all your local projects.
+
+### Reduce amount of generated debug information
+
+Recommendation: Add to your `Cargo.toml` or `.cargo/config.toml`:
+
+```toml
+[profile.dev]
+debug = "line-tables-only"
+
+[profile.dev.package."*"]
+debug = false
+
+[profile.debugging]
+inherits = "dev"
+debug = true
+```
+
+This will:
+- Change the [`dev` profile](../reference/profiles.md#dev) (default for development commands) to:
+  - Limit [debug information](../reference/profiles.md#debug) for workspace members to what is needed for useful panic backtraces
+  - Avoid generating any debug information for dependencies
+- Provide an opt-in for when debugging via [`--profile debugging`](../reference/profiles.md#custom-profiles)
+
+Trade-offs:
+- ✅ Faster build times
+- ✅ Faster link times
+- ✅ Smaller disk usage of the `target` directory
+- ❌ Requires a full rebuild to have a high-quality debugger experience
+
+### Use an alternative codegen backend
+
+> **This requires nightly/unstable features**
+
+Recommendation:
+
+- Install the Cranelift codegen backend rustup component
+    ```console
+    $ rustup component add rustc-codegen-cranelift-preview --toolchain nightly
+    ```
+- Add to your `Cargo.toml` or `.cargo/config.toml`:
+    ```toml
+    [profile.dev]
+    codegen-backend = "cranelift"
+    ```
+- Run Cargo with `-Z codegen-backend` or enable the [`codegen-backend`](../reference/unstable.md#codegen-backend) feature in `.cargo/config.toml`.
+  - This is required because this is currently an unstable feature.
+
+This will change the [`dev` profile](../reference/profiles.md#dev) to use the [Cranelift codegen backend](https://github.com/rust-lang/rustc_codegen_cranelift) for generating machine code, instead of the default LLVM backend. The Cranelift backend should generate code faster than LLVM, which should result in improved build performance.
+
+Trade-offs:
+- ✅ Faster build times
+- ❌ Worse runtime performance of the generated code
```

Comment:
If I'm understanding correctly, being more specific, this is
- This always speeds up builds of proc macros and build scripts
- For everything else, this only helps with `cargo build` and not `cargo check`
- `cargo build` will mostly be done through `cargo test` and this could make `cargo test` times worse than the savings in `cargo build`

Any way we could provide better nuance here?  Saying "faster build times" may sound more broad than it is.  And saying worse runtime performance, people may not pick up on the implications.